Added serialize and unserialize bug fixes

stephenwf opened the pull request:
@thewilkybarkid

thewilkybarkid commented on src/Serializer/SubjectsAware.php:
```diff
@@ -21,6 +22,9 @@
 
     final private function getSubjects(array $subjects) : PromiseInterface
     {
+        if ($subjects instanceof ArrayCollection) {
```
Typehint says no...

thewilkybarkid commented:
I think `Collection` should extend `JsonSerializable` directly.

stephenwf commented:
Hi @thewilkybarkid ,

I've added in the test and broken down what it does. Basically mimicking what our use case is (transporting as a string without losing data.) I did it in the format `[ string -> object -> string ]` to minimise the tests footprint and potentially allowing it to be fed by a data provider of troublesome JSON elements, instead of `[ object -> string -> object ]`. (Also if that particular step fails, it will be caught by the normalisation tests.)

Also not seen here is the addition of `ProvidesSubjects` interface. I added this to the ApiClient source without realising that its an external dependency. But its basically this:

``` php
<?php

namespace eLife\ApiClient\ApiClient;

use GuzzleHttp\Promise\PromiseInterface;

interface ProvidesSubjects {

  public function listSubjects(array $headers = [], int $page = 1, int $perPage = 20, bool $descendingOrder = TRUE) : PromiseInterface;

  public function getSubject(array $headers, string $id) : PromiseInterface;
}

```

and the SubjectsClient implements it. In a perfect world, this interface would live in a common interface repository to allow other implementations of clients (In search's case DoctrineCacheClient, which prevents the spamification to the API).

So the build has failed for now, and I'll leave it to your judgement on how to tackle the problems I mentioned in this PR (custom API clients, ArrayCollection casting).

Thanks,
Stephen

thewilkybarkid commented:
I see, so `ProvidesSubjects` is purely there to allow for caching (across processes)?

I did try to think about interfaces when developing the API client, but decided against it as we would have to break BC every time we changed anything. The clients, though, take a `HttpClient` where caching could be implemented (though a quicker method would be to use the [guzzle-cache-middleware](https://github.com/Kevinrob/guzzle-cache-middleware)).

stephenwf commented:
Thats a very good point, I guess that would abstract away the change. Okay Ill fix up this PR, and get that changed in the search impl.

stephenwf commented:
@thewilkybarkid 

Okay I've removed those lines. Are you okay with the normalizer test and the use case?

thewilkybarkid commented on src/Serializer/SubjectsAware.php:
```diff
@@ -19,8 +20,11 @@
     }
 
-    final private function getSubjects(array $subjects) : PromiseInterface
+    final private function getSubjects($subjects) : PromiseInterface
```
These changes aren't needed anymore are they?

stephenwf replied:
Unfortunately the `object -> json -> object` still uses ArrayCollection interchangeably

thewilkybarkid replied:
Reading the Symfony docs again, normalization is supposed to produce purely an array, so we're incorrect in leaving `Collection`s in there. So I think the _real_ fix for all of this is to either use `->toArray()` when normalizing, or do a `$this->normalizer->normalize($collection)`.

stephenwf replied:
So like this: https://github.com/elifesciences/api-sdk-php/pull/20/commits/42b354b5343d1d69d878a465accf91d1675351b5 ?

thewilkybarkid commented on test/Serializer/BlogArticleNormalizerTest.php:
```diff
@@ -78,6 +81,31 @@ public function canNormalizeProvider() : array
     /**
      * @test
+     */
+    public function is_can_serialize_and_deserialize_blog_articles()
```
I'd change the `ApiSdk` tests to test `serialize()` rather than `normalize()` instead of this (`BlogArticleNormalizer` only (de)normalizes, not (de)serialize).

stephenwf commented:
@thewilkybarkid Moved those tests to the API sdk now

stephenwf commented:
@thewilkybarkid If I can get this merged in, I will update the other entities with the toArray calls in a subsequent PR (to unblock Ross)

thewilkybarkid commented on src/Serializer/BlogArticleNormalizer.php:
```diff
@@ -66,13 +66,13 @@ public function normalize($object, $format = null, array $context = []) : array
+            })->toArray();
         }
 
         return all($data)->wait();
```
Use of `all()` is redundant now.

thewilkybarkid commented on test/Serializer/BlogArticleNormalizerTest.php:
```diff
@@ -41,7 +42,7 @@ protected function setUpNormalizer()
             new SubjectNormalizer(),
             new Block\ParagraphNormalizer(),
-        ]);
+        ], [new JsonEncoder()]);
```
Don't think this is needed.

thewilkybarkid commented:
I suggest we undo everything bar the `->toArray()` changes, and add it for other the other types here. We don't need the test either, as `JsonEncoder()` will always be able to turn only arrays into JSON.

stephenwf commented:
Cool, was easier to just redo in a new branch #22
